atomicfu/src/appleMain/kotlin/kotlinx/atomicfu/locks/Synchronized.kt

```diff
@@ -0,0 +1,297 @@
+            reEnterCount += 1
+            true
+        } else if (waiters.incrementAndGet() == 1 && owner.compareAndSet(NO_OWNER, self)) {
+            true
+        } else {
+            waiters.decrementAndGet()
+            false
+        }
+    }
+
+
+    private fun waitForUnlockAndLock(self: Long) {
+        withMonitor(monitor) {
+            while (!owner.compareAndSet(NO_OWNER, self)) {
+                monitor.waitWithDonation(owner.value)
+            }
+        }
+    }
+
+    public fun unlock() {
+        require (owner.value == currentThreadId)
+        if (reEnterCount > 0) {
+            reEnterCount -= 1
+        } else {
+            owner.value = NO_OWNER
+            if (waiters.decrementAndGet() > 0) {
+                withMonitor(monitor) {
+                    // We expect the highest priority thread to be woken up, but this should work
+                    // in any case.
+                    monitor.nativeMutex.notify()
+                }
+            }
+        }
+    }
+
+    private inline fun withMonitor(monitor: DonatingMonitor, block: () -> Unit) {
+        monitor.nativeMutex.enter()
+        return try {
+            block()
+        } finally {
+            monitor.nativeMutex.exit()
+        }
+    }
+
+    @OptIn(kotlin.experimental.ExperimentalNativeApi::class)
+    private class DonatingMonitor {
+        val nativeMutex = mutexPool.allocate()
+        val cleaner = createCleaner(nativeMutex) { mutexPool.release(it) }
+        var qosOverride: pthread_override_t? = null
+        var qosOverrideQosClass: UInt = 0U
+
+        fun waitWithDonation(lockOwner: Long) {
+            donateQos(lockOwner)
+            nativeMutex.wait()
+            clearDonation()
+        }
+
+        private fun donateQos(lockOwner: Long) {
+            if (lockOwner == NO_OWNER) {
+                return
+            }
+            val ourQosClass = qos_class_self() as UInt
+            // Set up a new override if required:
+            if (qosOverride != null) {
+                // There is an existing override, but we need to go higher.
+                if (ourQosClass > qosOverrideQosClass) {
+                    pthread_override_qos_class_end_np(qosOverride)
+                    qosOverride = pthread_override_qos_class_start_np(lockOwner.toCPointer(), qos_class_self(), 0)
+                    qosOverrideQosClass = ourQosClass
+                }
+            } else {
+                // No existing override, check if we need to set one up.
+                memScoped {
+                    val lockOwnerQosClass = alloc<UIntVar>()
+                    val lockOwnerRelPrio = alloc<IntVar>()
+                    pthread_get_qos_class_np(lockOwner.toCPointer(), lockOwnerQosClass.ptr, lockOwnerRelPrio.ptr)
+                    if (ourQosClass > lockOwnerQosClass.value) {
+                        qosOverride = pthread_override_qos_class_start_np(lockOwner.toCPointer(), qos_class_self(), 0)
```

Comment:
I think there is a race condition possible here:
Let’s say we have thread 1 and 2 with low priority and thread 3 with high priority. 

Thread 1 is holding the lock and doing work, thread 2 is suspended and waiting for the lock to be released. There is no qos override since 1 and 2 have equal priority. 

Now, if thread 3 (high priority) tries to acquire the lock it would reach this point (line 173) in the code (please correct me if I’m wrong). If in the meantime thread 1 releases the lock and thread 2 therefore acquires it, then the `lockOwner` reference is outdated and we would override priority from thread 1 and not the new lock holder thread 2.

Reply:
Thread 2 shouldn't be able to acquire the mutex in waitForUnlockAndLock while Thread 3 holds it on line 173?

Reply:
Ah I see, donation happens between pthread_mutex_lock and pthread_mutex_wait. My bad!